### DISCUSSION TITLE
String Concat / Template literals not working. Tailwinds 3.0

### DISCUSSION THREAD

**Discussion by codebytesfl** [tailwindlabs/tailwindcss#6745] [Help] [answered]
**Node**: 16.3.0
**PostCSS**: 8.4.5
**React**: 16.14.0
**Tailwind**: 3.0.7

Seems I found a weird bug when using string concat / template literals. Am I missing something here or is this a bug.

`tailwind.config.js`
```javascript
const colors = require('tailwindcss/colors');

module.exports = {
    content: ['./src/**/*.{js,ts,jsx,tsx}'],
    theme: {
        extend: {},
        colors: {
            primary: { DEFAULT: '#00a5a8', hover: '#007275', active: '#00595c', disabled: '#b2e4e5' },
        },
    },
    plugins: [],
};

```

**DOES NOT WORK**

`button.tsx`
```javascript
const Button: React.FC<ButtonProps> = ({ children, onClick, variant = 'primary', }) => {

    let test = 'primary';
    let background = `bg-${test}`;

    return (
        <button onClick={onClick} data-testid="Button" className={background}>
            {children}
        </button>
    );
};
```

**ALSO DOES NOT WORK**

`button.tsx`
```javascript
const Button: React.FC<ButtonProps> = ({ children, onClick, variant = 'primary', }) => {

    let test = 'primary';
    let background = 'bg-' + test;

    return (
        <button onClick={onClick} data-testid="Button" className={background}>
            {children}
        </button>
    );
};
```

**ALSO DOES NOT WORK**

`button.tsx`
```javascript
const Button: React.FC<ButtonProps> = ({ children, onClick, variant = 'primary', }) => {
    let background = 'bg-' + 'primary';

    return (
        <button onClick={onClick} data-testid="Button" className={background}>
            {children}
        </button>
    );
};
```

**DOES WORK**

`button.tsx`
```javascript
const Button: React.FC<ButtonProps> = ({ children, onClick, variant = 'primary', }) => {

    let background = `bg-primary`;

    return (
        <button onClick={onClick} data-testid="Button" className={background}>
            {children}
        </button>
    );
};
```

`postcss.config.js`
```javascript
module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
  },
};
```

**Comment by codebytesfl** [comment]
Well, some lovely googling revealed that string concat is not a good idea in tailwinds. Back to the drawing board I go! The idea is that users could have a `n` amount of colors in their tailwind config and just pass whatever color they want to the button component and it would render the right colors.

**Comment by adamwathan** [comment, marked as answer]
Here’s the relevant documentation on this in case anyone finds this thread with the same problem:

https://tailwindcss.com/docs/content-configuration#dynamic-class-names

**Reply by bonnie** [reply]
Template literals worked for me in 2.x, broke with 3.x (though if I'd read the docs I'd know I was treading on thin ice: https://v2.tailwindcss.com/docs/just-in-time-mode#arbitrary-value-support). Twind (https://twind.dev/) looks like it will be useful to me for TailwindCSS v3.

**Reply by codebytesfl** [reply]
@bonnie  That is what threw me off. It worked rather well for me in 2.x but I wanted to try 3.x. Twind looks to be a solid solution to look into. Thanks for sharing that!

**Comment by usr-ein** [comment]
I know this is not the most satisfying solution, but since Tailwind JIT/Postcss run some sort of primitive logic on the JSX code, a workaround I've found is to generate all the classes you dynamically want to use, and write them into an invisible component's `className`, like so:
```jsx
<div className={bg-${color}}>Hehehehehe</div>
<span
    className="hidden
            bg-success hover:bg-success bg-success-darker hover:bg-success-darker
            bg-info hover:bg-info bg-info-darker hover:bg-info-darker
            bg-warning hover:bg-warning bg-warning-darker hover:bg-warning-darker
            bg-danger hover:bg-danger bg-danger-darker hover:bg-danger-darker
            ">
                a
            </span>
```
Then, on transpillation, Tailwind & PostCSS will think the classes need to be generated, and will include them in the build, which means they'll be available for the dynamicly computed classNames too!

**Reply by Ross-mc** [reply]
A much better solution as listed in the official docs is to [Safelist](https://tailwindcss.com/docs/content-configuration#safelisting-classes) classes to ensure that they are always generated, this way you don't need to pollute your mark up with an invisible component.

**Comment by Temez1** [comment]
> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as complete unbroken strings in your source files.
> 
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS

Could this be mentioned somewhere earlier in the docs? I was using Astro & spent 2 hours to debug & reproducing and try to understand the problem, because I thought it had something to do with Astro. Turned out the reason why some of it was working was because the class was used at other part of the program. The confusing part is, that they are applied to the html classes code, but not the css classes. Is there a reason why they are still left in the classes but just not applied?
